export.md

```diff
@@ -1,4 +1,14 @@
 # FHIR Bulk Data Export API Proposal
+ 
+## Security Considerations
+
+When initiating any request with the FHIR resource server, the client first MUST use [Transport Layer Security (TLS) Protocol Version 1.2 (RFC5246)](https://tools.ietf.org/html/rfc5246) to authenticate the identity of the FHIR server and to establish an encrypted, integrity-protected link for securing all exchanges between the client and the server involved in the exchange.  All exchanges described herein between a client and a server MUST be secured using TLS V1.2.  Use of mutual TLS is OPTIONAL.  
+
+Prior to kicking off a bulk-data request, the authorization server responsible for authorizing access to the desired FHIR resources MUST have authenticated the client and issued it one or more access tokens for use in requesting the export of FHIR resources, status information, and file downloads (which could involve multiple servers).  For this purpose, the OAuth 2.0 based SMART Backend Services: Authorization Guide is RECOMMENDED.
+
+The decision whether to issue a single access token for use for all three purposes, or to issue multiple tokens is a risk-management decision to be made by the organization responsible for managing access to the resources. To minimize risks associated with token redirection, the scope of each access token SHOULD encompass, and be limited to, the resources requested.  Also, in some cases, the server that receives status queries may be different from the FHIR resource server.  So the use of separate access tokens for requesting FHIR resources and for requesting status information is RECOMMENDED. The scope of resources applicable to the access token required (optionally) for file download will be identical to (or a subset of) the token used for the initial request.  So the authorization server MAY choose to allow the use of the same token for both purposes, or it MAY choose to issue a new token at the time needed for file download.
```

Comment:
I'm a bit concerned about this recommendation if we don't also provide a language and distinct scopes to partition off these different capabilities -- in other words unless the client has some way to ask for "a token that can do a kickoff" vs "a token that can poll for status updates" vs "a token that can fetch results files", this scheme doesn't really work.

So we'd need to specify a fair bit more to make this work. Given that the client's authentication method for all three of these is identical, and given that the tokens are short-lived, I'm not sure what threat we'd be mitigating. Can we be clear on this @bakerdb? (My current position is that I'd rather avoid adding this complexity, based on what I know right now.)

Reply:
@jmandel  The endpoint for a status query may be different from the endpoint for the data request, and the endpoint for file retrieval may be different from either of these endpoints.  (In particular, Grahame pointed this out on the Zulip discussion.)  Also, the time required for a bulk-data export may be quite long (e.g., for genomic data, image data, or system backup).  So I don't think it's reasonable to expect that a token issued for the request, and with an expiration no greater than 5 minutes, would be valid across multiple endpoints and  long periods of time.  At the same time, policy among implementers will differ based on data types, computing power, risk tolerance, etc.  So I think we should raise this topic, but would welcome further discussion.

Reply:
In reviewing the Bulk Data Implementation Guide, I noticed that the Response - Complete Status section says "requiresAccessToken - boolean value indicating whether downloading the generated files will require an AUTHENTICATION TOKEN. Note: This may be false in the case of signed S3 URIs or an internal file server within an organization's firewall." [all-caps added]  But the File Request section says "If the requiresAccessToken field in the Complete Status body is set to true, the request MUST include a valid ACCESS TOKEN in the Authorization header (i.e., Authorization: Bearer {{token}})."  Accessing the generated files is arguably the most sensitive step in the bulk-data-access scenario, as it is the only step that can provide direct access to PHI.  IMO, an access token should be REQUIRED, with the authentication JWT optional (for the reason stated in the Response - Complete section).  I also think this access token should be different from that used to request extraction, because if the same token is used for both purposes, then if an adversary captures the "request" token, the worst damage is potential lost cycles for generating the files.  But if that adversary can then use the same token to download the files, the risk is significant.

Reply:
Thanks @bakerdb  -- we should fix the phrase "authentication token" to "access token" to be consistent with our own definitions.

Beyond this, I want to highlight that the current scope system we define is based on SMART's resource-oriented scopes, so a client requests a token with a scope like `system/*.read` rather than `bulk-data-kickoff` or `bulk-data-status-check` or `bulk-data-download-results`. This `system/*.read` scope allows the client to participate in each step of the bulk export process, just as it allows for non-bulk (i.e., targeted) export through queries like` GET /Observation?patient=123`. In other words, the current bulk data access spec builds on the existing scope language and permission system, and unifies the authorization process for sync+async access, which I think is a good thing.

In terms of the threat model, do we have a reason to think that different factors contribute to adversaries observing the kick-off phase vs the download phase? I think one of the outcomes from our last phone discussion was to try to highlight what these factors would be.